File: pkg/adf/build_test.go
// SPDX-FileCopyrightText: (c) 2026 Rafal Zajac
// SPDX-License-Identifier: MIT

package adf

import (
	"testing"

	"github.com/ctx42/testing/pkg/assert"
	"github.com/ctx42/testing/pkg/must"
)

// emptyDoc is a baseline with no content, the shape a page create starts from.
const emptyDoc = `{ "adf": { "type": "doc", "content": [] } }`

func Test_ADF_Put_insertRoundTrip_tabular(t *testing.T) {
	tt := []struct {
		testN    string
		body     string
		wantType string
	}{
		{
			"toc marker",
			"[[TOC]]",
			"extension",
		},
		{
			"code block with language",
			"```go\nx := 1\n```",
			"codeBlock",
		},
		{
			"code block without language",
			"```\nplain\n```",
			"codeBlock",
		},
		{
			"code block with blank lines and pipes",
			"```plaintext\n| a |\n\n| b |\n```",
			"codeBlock",
		},
		{
			"bullet list",
			"- one\n- two",
			"bulletList",
		},
		{
			"bullet list with inline formatting",
			"- `ID` is a **code**\n- plain",
			"bulletList",
		},
		{
			"ordered list",
			"1. one\n2. two",
			"orderedList",
		},
		{
			"ordered list starting past one",
			"3. three\n4. four",
			"orderedList",
		},
		{
			"blockquote",
			"> hello",
			"blockquote",
		},
		{
			"blockquote with two paragraphs",
			"> alpha\n>\n> beta",
			"blockquote",
		},
		{
			"note panel",
			"> [!NOTE]\n> body",
			"panel",
		},
		{
			"warning panel",
			"> [!WARNING]\n> body",
			"panel",
		},
		{
			"expand with title",
			"> [!EXPAND] More detail\n> body",
			"expand",
		},
		{
			"expand without title",
			"> [!EXPAND]\n> body",
			"expand",
		},
		{
			"table with header row",
			"| a | b |\n|---|---|\n| 1 | 2 |",
			"table",
		},
		{
			"table without header row",
			"|   |   |\n|---|---|\n| k | v |",
			"table",
		},
		{
			"table with escaped pipe",
			"| a \\| b |\n|--------|\n| c      |",
			"table",
		},
	}
	for _, tc := range tt {
		t.Run(tc.testN, func(t *testing.T) {
			// --- Given ---
			base := must.Value(NewADF([]byte(emptyDoc)))

			// --- When ---
			out := must.Value(base.Put(tc.body, nil, nil, nil))

			// --- Then ---
			assert.Equal(t, tc.body, renderBody(t, out, nil))
			assert.Equal(t, tc.wantType, out.Doc.Content[0].Type)
		})
	}
}

func Test_ADF_Put_insertToc(t *testing.T) {
	// --- Given ---
	base := must.Value(NewADF([]byte(emptyDoc)))

	// --- When ---
	out := must.Value(base.Put("[[TOC]]", nil, nil, nil))

	// --- Then ---
	node := out.Doc.Content[0]
	assert.Equal(t, "extension", node.Type)
	assert.Equal(t, "toc", node.attrStr("extensionKey"))
	want := "com.atlassian.confluence.macro.core"
	assert.Equal(t, want, node.attrStr("extensionType"))
	assert.Equal(t, "default", node.attrStr("layout"))
}

func Test_ADF_Put_insertCodeBlock(t *testing.T) {
	t.Run("language and body", func(t *testing.T) {
		// --- Given ---
		base := must.Value(NewADF([]byte(emptyDoc)))

		// --- When ---
		out := must.Value(base.Put("```go\nx := 1\ny := 2\n```", nil, nil, nil))

		// --- Then ---
		node := out.Doc.Content[0]
		assert.Equal(t, "codeBlock", node.Type)
		assert.Equal(t, "go", node.attrStr("language"))
		assert.Equal(t, "x := 1\ny := 2", node.codeText())
	})

	t.Run("empty body", func(t *testing.T) {
		// --- Given ---
		base := must.Value(NewADF([]byte(emptyDoc)))

		// --- When ---
		out := must.Value(base.Put("```\n\n```", nil, nil, nil))

		// --- Then ---
		node := out.Doc.Content[0]
		assert.Equal(t, "codeBlock", node.Type)
		assert.Equal(t, "", node.codeText())
	})
}

func Test_ADF_Put_insertBulletList(t *testing.T) {
	// --- Given ---
	base := must.Value(NewADF([]byte(emptyDoc)))
	body := "- one\n- two wrapped\n  over lines"

	// --- When ---
	out := must.Value(base.Put(body, nil, nil, nil))

	// --- Then ---
	node := out.Doc.Content[0]
	assert.Equal(t, "bulletList", node.Type)
	assert.Equal(t, 2, len(node.Content))
	assert.Equal(t, "listItem", node.Content[0].Type)
	assert.Equal(t, "one", node.Content[0].Content[0].Content[0].Text)
	assert.Equal(t, "listItem", node.Content[1].Type)
}

func Test_ADF_Put_insertOrderedList(t *testing.T) {
	t.Run("starting at one omits the order attribute", func(t *testing.T) {
		// --- Given ---
		base := must.Value(NewADF([]byte(emptyDoc)))

		// --- When ---
		out := must.Value(base.Put("1. one\n2. two", nil, nil, nil))

		// --- Then ---
		node := out.Doc.Content[0]
		assert.Equal(t, "orderedList", node.Type)
		assert.Equal(t, 2, len(node.Content))
		assert.Nil(t, node.Attrs)
	})

	t.Run("starting past one records the start", func(t *testing.T) {
		// --- Given ---
		base := must.Value(NewADF([]byte(emptyDoc)))

		// --- When ---
		out := must.Value(base.Put("3. three\n4. four", nil, nil, nil))

		// --- Then ---
		node := out.Doc.Content[0]
		assert.Equal(t, "orderedList", node.Type)
		assert.Equal(t, 3, node.attrInt("order"))
	})
}

func Test_ADF_Put_insertPanel(t *testing.T) {
	// --- Given ---
	base := must.Value(NewADF([]byte(emptyDoc)))

	// --- When ---
	out := must.Value(base.Put("> [!SUCCESS]\n> alpha\n>\n> beta", nil, nil, nil))

	// --- Then ---
	node := out.Doc.Content[0]
	assert.Equal(t, "panel", node.Type)
	assert.Equal(t, "success", node.attrStr("panelType"))
	assert.Equal(t, 2, len(node.Content))
	assert.Equal(t, "paragraph", node.Content[0].Type)
	assert.Equal(t, "alpha", node.Content[0].Content[0].Text)
}

func Test_ADF_Put_insertExpand(t *testing.T) {
	t.Run("with title", func(t *testing.T) {
		// --- Given ---
		base := must.Value(NewADF([]byte(emptyDoc)))

		// --- When ---
		out := must.Value(base.Put("> [!EXPAND] More\n> body", nil, nil, nil))

		// --- Then ---
		node := out.Doc.Content[0]
		assert.Equal(t, "expand", node.Type)
		assert.Equal(t, "More", node.attrStr("title"))
	})

	t.Run("without title", func(t *testing.T) {
		// --- Given ---
		base := must.Value(NewADF([]byte(emptyDoc)))

		// --- When ---
		out := must.Value(base.Put("> [!EXPAND]\n> body", nil, nil, nil))

		// --- Then ---
		node := out.Doc.Content[0]
		assert.Equal(t, "expand", node.Type)
		assert.Nil(t, node.Attrs)
	})
}

func Test_ADF_Put_insertTable(t *testing.T) {
	t.Run("header row becomes tableHeader cells", func(t *testing.T) {
		// --- Given ---
		base := must.Value(NewADF([]byte(emptyDoc)))
		body := "| a | b |\n|---|---|\n| 1 | 2 |"

		// --- When ---
		out := must.Value(base.Put(body, nil, nil, nil))

		// --- Then ---
		node := out.Doc.Content[0]
		assert.Equal(t, "table", node.Type)
		assert.Equal(t, 2, len(node.Content))
		assert.Equal(t, "tableHeader", node.Content[0].Content[0].Type)
		assert.Equal(t, "tableHeader", node.Content[0].Content[1].Type)
		assert.Equal(t, "tableCell", node.Content[1].Content[0].Type)
		cell := node.Content[1].Content[0]
		assert.Equal(t, "1", cell.Content[0].Content[0].Text)
	})

	t.Run("blank header row yields no header cells", func(t *testing.T) {
		// --- Given ---
		base := must.Value(NewADF([]byte(emptyDoc)))
		body := "|   |   |\n|---|---|\n| k | v |"

		// --- When ---
		out := must.Value(base.Put(body, nil, nil, nil))

		// --- Then ---
		node := out.Doc.Content[0]
		assert.Equal(t, 1, len(node.Content))
		assert.Equal(t, "tableCell", node.Content[0].Content[0].Type)
	})

	t.Run("br splits a cell into paragraphs", func(t *testing.T) {
		// --- Given ---
		base := must.Value(NewADF([]byte(emptyDoc)))
		body := "| h      |\n|--------|\n| a<br>b |"

		// --- When ---
		out := must.Value(base.Put(body, nil, nil, nil))

		// --- Then ---
		cell := out.Doc.Content[0].Content[1].Content[0]
		assert.Equal(t, 2, len(cell.Content))
		assert.Equal(t, "a", cell.Content[0].Content[0].Text)
		assert.Equal(t, "b", cell.Content[1].Content[0].Text)
	})

	t.Run("escaped pipe stays inside its cell", func(t *testing.T) {
		// --- Given ---
		base := must.Value(NewADF([]byte(emptyDoc)))
		body := "| a \\| b |\n|--------|\n| c      |"

		// --- When ---
		out := must.Value(base.Put(body, nil, nil, nil))

		// --- Then ---
		cell := out.Doc.Content[0].Content[0].Content[0]
		assert.Equal(t, "a | b", cell.Content[0].Content[0].Text)
	})
}

func Test_ADF_Put_insertIntoExisting(t *testing.T) {
	// --- Given ---
	data := `{ "adf": { "type": "doc", "content": [
	   { "type": "paragraph", "attrs": { "localId": "p1" },
	     "content": [ { "type": "text", "text": "alpha" } ] } ] } }`
	base := must.Value(NewADF([]byte(data)))
	body := "alpha\n\n- one\n- two"

	// --- When ---
	out := must.Value(base.Put(body, nil, nil, nil))

	// --- Then ---
	assert.Equal(t, body, renderBody(t, out, nil))
	assert.Equal(t, "p1", out.Doc.Content[0].attrStr("localId"))
	assert.Equal(t, "bulletList", out.Doc.Content[1].Type)
}

func Test_ADF_Put_insertRejects_tabular(t *testing.T) {
	tt := []struct {
		testN   string
		body    string
		wantErr string
	}{
		{
			"star bullet marker",
			"* one\n* two",
			"write bullet items with a \"- \" marker",
		},
		{
			"plus bullet marker",
			"+ one",
			"write bullet items with a \"- \" marker",
		},
		{
			"nested list in a bullet list",
			"- one\n  - nested",
			"a nested block cannot be inserted",
		},
		{
			"nested list in an ordered list",
			"1. one\n   - nested",
			"a nested block cannot be inserted",
		},
		{
			"multi-paragraph list item",
			"- one\n\n  two",
			"only single-paragraph",
		},
		{
			"non-sequential ordered list",
			"1. one\n3. three",
			"items must be numbered sequentially",
		},
		{
			"table without separator row",
			"| a | b |",
			"needs a header and a separator row",
		},
		{
			"table with a malformed separator row",
			"| a | b |\n| x | y |\n| 1 | 2 |",
			"missing its '---' separator row",
		},
		{
			"table with ragged rows",
			"| a | b |\n|---|---|\n| 1 |",
			"every table row needs 2 cells",
		},
		{
			"table with a span marker",
			"| a | b |\n|---|---|\n| « | 2 |",
			"cell spans cannot be inserted",
		},
		{
			"unknown panel tag",
			"> [!BOGUS]\n> body",
			"unknown panel type",
		},
		{
			"custom panel tag",
			"> [!CUSTOM]\n> body",
			"unknown panel type",
		},
		{
			"panel without a body",
			"> [!NOTE]",
			"needs a body",
		},
		{
			"nested block in a blockquote",
			"> - item",
			"a nested block cannot be inserted",
		},
		{
			"toc marker with trailing text",
			"[[TOC]] extra",
			"cannot insert block",
		},
		{
			"anchor comment",
			"<!-- adf:unsupported -->",
			"cannot insert block",
		},
		{
			"image with no uploaded attachment",
			"![alt](missing.png)",
			"no uploaded attachment",
		},
	}
	for _, tc := range tt {
		t.Run(tc.testN, func(t *testing.T) {
			// --- Given ---
			base := must.Value(NewADF([]byte(emptyDoc)))

			// --- When ---
			_, err := base.Put(tc.body, nil, nil, nil)

			// --- Then ---
			assert.ErrorContain(t, tc.wantErr, err)
		})
	}
}

File: packages/core/test/adf/lens/build.test.ts
// SPDX-FileCopyrightText: (c) 2026 Rafal Zajac
// SPDX-License-Identifier: MIT

// Ported from pkg/adf/build_test.go — the insert builders driven through `put`.
// Each case inserts a block into an empty document and asserts it round-trips:
// `renderBody(out)` equals the inserted body (the PutGet lens law, enforced by
// validatePut) and the built node has the expected type. This is the M4.4
// validation coverage for inserts; the reject table is the "frozen/lossy edit is
// refused with a named reason" half.
//
// Dialect divergences from Go: an image inserts as an Obsidian embed `![[…]]`
// (not `![alt](path)`); the `[[TOC]]` marker is evicted, so there is no toc-macro
// insert (a frozen ```adf block round-trips losslessly as a read-only code
// block). Everything else is dialect-stable and ports 1:1.

import { describe, expect, it } from "vitest";
import { put } from "../../../src/adf/lens/reconstruct.ts";
import { marshallMarkdownMapped } from "../../../src/index.ts";
import { type ADF, attrInt, attrStr, newADF } from "../../../src/models/adf.ts";

// emptyDoc is a baseline with no content, the shape a page create starts from.
const emptyDoc = `{ "adf": { "type": "doc", "content": [] } }`;

/** renderBody renders adf and returns its body without frontmatter or the trailing newline. */
function renderBody(adf: ADF): string {
    const [md, sm] = marshallMarkdownMapped(adf, {});
    return md.slice(sm.bodyStart).replace(/\n$/, "");
}

/** insert puts body into an empty document and returns the rebuilt document. */
const insert = (body: string): ADF =>
    put(newADF(emptyDoc), body, null, null, null);

describe("put insert round-trip (tabular)", () => {
    const cases: Array<{ name: string; body: string; wantType: string }> = [
        {
            name: "code block with language",
            body: "```go\nx := 1\n```",
            wantType: "codeBlock",
        },
        {
            name: "code block without language",
            body: "```\nplain\n```",
            wantType: "codeBlock",
        },
        {
            name: "code block with blank lines and pipes",
            body: "```plaintext\n| a |\n\n| b |\n```",
            wantType: "codeBlock",
        },
        { name: "bullet list", body: "- one\n- two", wantType: "bulletList" },
        {
            name: "bullet list with inline formatting",
            body: "- `ID` is a **code**\n- plain",
            wantType: "bulletList",
        },
        {
            name: "ordered list",
            body: "1. one\n2. two",
            wantType: "orderedList",
        },
        {
            name: "ordered list starting past one",
            body: "3. three\n4. four",
            wantType: "orderedList",
        },
        { name: "blockquote", body: "> hello", wantType: "blockquote" },
        {
            name: "blockquote with two paragraphs",
            body: "> alpha\n>\n> beta",
            wantType: "blockquote",
        },
        { name: "note panel", body: "> [!NOTE]\n> body", wantType: "panel" },
        {
            name: "warning panel",
            body: "> [!WARNING]\n> body",
            wantType: "panel",
        },
        {
            name: "expand with title",
            body: "> [!EXPAND] More detail\n> body",
            wantType: "expand",
        },
        {
            name: "expand without title",
            body: "> [!EXPAND]\n> body",
            wantType: "expand",
        },
        {
            name: "table with header row",
            body: "| a | b |\n|---|---|\n| 1 | 2 |",
            wantType: "table",
        },
        {
            name: "table without header row",
            body: "|   |   |\n|---|---|\n| k | v |",
            wantType: "table",
        },
        {
            name: "table with escaped pipe",
            body: "| a \\| b |\n|--------|\n| c      |",
            wantType: "table",
        },
    ];

    for (const tc of cases) {
        it(tc.name, () => {
            const out = insert(tc.body);
            expect(renderBody(out)).toBe(tc.body);
            expect(out.doc.content?.[0]?.type).toBe(tc.wantType);
        });
    }
});

describe("put insertCodeBlock", () => {
    it("language and body", () => {
        const node = insert("```go\nx := 1\ny := 2\n```").doc.content?.[0];
        expect(node?.type).toBe("codeBlock");
        expect(attrStr(node?.attrs, "language")).toBe("go");
        expect(node?.content?.[0]?.text).toBe("x := 1\ny := 2");
    });

    it("empty body", () => {
        const node = insert("```\n\n```").doc.content?.[0];
        expect(node?.type).toBe("codeBlock");
        expect(node?.content ?? []).toEqual([]);
    });
});

describe("put insertTocMacro", () => {
    it("rebuilds an inserted TOC block into a live toc extension", () => {
        const node = insert("```adf\ntype: toc\nlocalId: e1\n```").doc
            .content?.[0];
        expect(node?.type).toBe("extension");
        expect(attrStr(node?.attrs, "extensionKey")).toBe("toc");
        expect(attrStr(node?.attrs, "extensionType")).toBe(
            "com.atlassian.confluence.macro.core",
        );
        expect(attrStr(node?.attrs, "localId")).toBe("e1");
    });

    it("rebuilds a TOC block that carries no localId", () => {
        const node = insert("```adf\ntype: toc\n```").doc.content?.[0];
        expect(node?.type).toBe("extension");
        expect(attrStr(node?.attrs, "extensionKey")).toBe("toc");
        expect(attrStr(node?.attrs, "localId")).toBe("");
    });

    it("keeps a non-toc frozen macro as a read-only code block", () => {
        const node = insert(
            "```adf\ntype: extension\nextensionKey: chart\nlocalId: e2\n```",
        ).doc.content?.[0];
        expect(node?.type).toBe("codeBlock");
        expect(attrStr(node?.attrs, "language")).toBe("adf");
    });
});

describe("put insertBulletList", () => {
    it("splits items and unwraps a soft-wrapped item", () => {
        const node = insert("- one\n- two wrapped\n  over lines").doc
            .content?.[0];
        expect(node?.type).toBe("bulletList");
        expect(node?.content?.length).toBe(2);
        expect(node?.content?.[0]?.type).toBe("listItem");
        expect(node?.content?.[0]?.content?.[0]?.content?.[0]?.text).toBe(
            "one",
        );
        expect(node?.content?.[1]?.type).toBe("listItem");
    });
});

describe("put insertOrderedList", () => {
    it("starting at one omits the order attribute", () => {
        const node = insert("1. one\n2. two").doc.content?.[0];
        expect(node?.type).toBe("orderedList");
        expect(node?.content?.length).toBe(2);
        expect(node?.attrs).toBeUndefined();
    });

    it("starting past one records the start", () => {
        const node = insert("3. three\n4. four").doc.content?.[0];
        expect(node?.type).toBe("orderedList");
        expect(attrInt(node?.attrs, "order")).toBe(3);
    });
});

describe("put insertPanel", () => {
    it("builds a typed panel with its body paragraphs", () => {
        const node = insert("> [!SUCCESS]\n> alpha\n>\n> beta").doc
            .content?.[0];
        expect(node?.type).toBe("panel");
        expect(attrStr(node?.attrs, "panelType")).toBe("success");
        expect(node?.content?.length).toBe(2);
        expect(node?.content?.[0]?.type).toBe("paragraph");
        expect(node?.content?.[0]?.content?.[0]?.text).toBe("alpha");
    });
});

describe("put insertExpand", () => {
    it("with title", () => {
        const node = insert("> [!EXPAND] More\n> body").doc.content?.[0];
        expect(node?.type).toBe("expand");
        expect(attrStr(node?.attrs, "title")).toBe("More");
    });

    it("without title", () => {
        const node = insert("> [!EXPAND]\n> body").doc.content?.[0];
        expect(node?.type).toBe("expand");
        expect(node?.attrs).toBeUndefined();
    });
});

describe("put insertTable", () => {
    it("header row becomes tableHeader cells", () => {
        const node = insert("| a | b |\n|---|---|\n| 1 | 2 |").doc.content?.[0];
        expect(node?.type).toBe("table");
        expect(node?.content?.length).toBe(2);
        expect(node?.content?.[0]?.content?.[0]?.type).toBe("tableHeader");
        expect(node?.content?.[0]?.content?.[1]?.type).toBe("tableHeader");
        expect(node?.content?.[1]?.content?.[0]?.type).toBe("tableCell");
        expect(
            node?.content?.[1]?.content?.[0]?.content?.[0]?.content?.[0]?.text,
        ).toBe("1");
    });

    it("blank header row yields no header cells", () => {
        const node = insert("|   |   |\n|---|---|\n| k | v |").doc.content?.[0];
        expect(node?.content?.length).toBe(1);
        expect(node?.content?.[0]?.content?.[0]?.type).toBe("tableCell");
    });

    it("br splits a cell into paragraphs", () => {
        const cell = insert("| h      |\n|--------|\n| a<br>b |").doc
            .content?.[0]?.content?.[1]?.content?.[0];
        expect(cell?.content?.length).toBe(2);
        expect(cell?.content?.[0]?.content?.[0]?.text).toBe("a");
        expect(cell?.content?.[1]?.content?.[0]?.text).toBe("b");
    });

    it("escaped pipe stays inside its cell", () => {
        const cell = insert("| a \\| b |\n|--------|\n| c      |").doc
            .content?.[0]?.content?.[0]?.content?.[0];
        expect(cell?.content?.[0]?.content?.[0]?.text).toBe("a | b");
    });
});

describe("put insertIntoExisting", () => {
    it("inserts a list after a kept paragraph", () => {
        const data = `{ "adf": { "type": "doc", "content": [
           { "type": "paragraph", "attrs": { "localId": "p1" },
             "content": [ { "type": "text", "text": "alpha" } ] } ] } }`;
        const body = "alpha\n\n- one\n- two";
        const out = put(newADF(data), body, null, null, null);
        expect(renderBody(out)).toBe(body);
        expect(attrStr(out.doc.content?.[0]?.attrs, "localId")).toBe("p1");
        expect(out.doc.content?.[1]?.type).toBe("bulletList");
    });
});

describe("put insert rejects (tabular)", () => {
    const cases: Array<{ name: string; body: string; want: string }> = [
        {
            name: "star bullet marker",
            body: "* one\n* two",
            want: 'write bullet items with a "- " marker',
        },
        {
            name: "plus bullet marker",
            body: "+ one",
            want: 'write bullet items with a "- " marker',
        },
        {
            name: "nested list in a bullet list",
            body: "- one\n  - nested",
            want: "a nested block cannot be inserted",
        },
        {
            name: "nested list in an ordered list",
            body: "1. one\n   - nested",
            want: "a nested block cannot be inserted",
        },
        {
            name: "multi-paragraph list item",
            body: "- one\n\n  two",
            want: "only single-paragraph",
        },
        {
            name: "non-sequential ordered list",
            body: "1. one\n3. three",
            want: "items must be numbered sequentially",
        },
        {
            name: "table without separator row",
            body: "| a | b |",
            want: "needs a header and a separator row",
        },
        {
            name: "table with a malformed separator row",
            body: "| a | b |\n| x | y |\n| 1 | 2 |",
            want: "missing its '---' separator row",
        },
        {
            name: "table with ragged rows",
            body: "| a | b |\n|---|---|\n| 1 |",
            want: "every table row needs 2 cells",
        },
        {
            name: "table with a span marker",
            body: "| a | b |\n|---|---|\n| « | 2 |",
            want: "cell spans cannot be inserted",
        },
        {
            name: "unknown panel tag",
            body: "> [!BOGUS]\n> body",
            want: "unknown panel type",
        },
        {
            name: "custom panel tag",
            body: "> [!CUSTOM]\n> body",
            want: "unknown panel type",
        },
        {
            name: "panel without a body",
            body: "> [!NOTE]",
            want: "needs a body",
        },
        {
            name: "nested block in a blockquote",
            body: "> - item",
            want: "a nested block cannot be inserted",
        },
        {
            name: "wikilink marker block",
            body: "[[TOC]] extra",
            want: "cannot insert block",
        },
        {
            name: "anchor comment",
            body: "<!-- adf:unsupported -->",
            want: "cannot insert block",
        },
        {
            name: "image with no uploaded attachment",
            body: "![[missing.png]]",
            want: "no uploaded attachment",
        },
    ];

    for (const tc of cases) {
        it(tc.name, () => {
            expect(() => insert(tc.body)).toThrow(tc.want);
        });
    }
});
